### PR TITLE
fix: improve performance

### PR DESCRIPTION
**Pull request by marcelltoth**
Performance improvements.

See stoplightio/elements#1064 for details.

1. `lastIndexOf` is surprisingly much slower than `indexOf`, like we're talking 35x slower. https://www.measurethat.net/Benchmarks/Show/3384/0/index-vs-lastindexof#latest_results_block This created a performance bottleneck when parsing large documents, with hundereds of refs. A more invasive but even better solution would be to just skip the document entirely if it's already bundled. 
2. $ref parsing is called many times and url encoding/decoding is slow. (For the stripe example I measured ~1s for that single function). When we only need the length, no need to parse the whole thing.

### CODE REVIEW

File: lib/pointer.js [hunk 1]
```diff
@@ -189,6 +189,25 @@ Pointer.parse = function (path, originalPath) {
   return pointer.slice(1);
 };
 
+
+/**
+ * Like `Pointer.parse` but only returns the length of the pointer.
```
Comment (P0lip) [inline lib/pointer.js:194]: Hmm, this naming is very confusing.
By `length` one could think of an actual length of a string, while what we actually verify here is the level / depth.

File: lib/bundle/index.js [hunk 2]
```diff
@@ -56,20 +56,33 @@ function crawl (parent, key, path, pathFromRoot, indirections, inventory, $refs,
...
-          let bDefinitionsIndex = `${pathFromRoot}/${b}`.lastIndexOf(options.bundle.defaultRoot);
-          // Most people will expect references to be bundled into the the defaultRoot property,
-          // so we always crawl that property first, if it exists.
-
-          if (aDefinitionsIndex !== bDefinitionsIndex) {
-            // Give higher priority to the $ref that's closer to the "definitions" property
-            return bDefinitionsIndex - aDefinitionsIndex;
+
+
+          // performance optimization: indexOf is ~12 times faster than lastIndexOf as of now, LIO becomes a bottleneck in larger schemas
```
Comment (P0lip) [inline lib/bundle/index.js:61]: hmmm, this might be tad too optimistic. I feel like this check may actually hamper the performance in other cases.

Apparently, $refs in Stripe doesn't match `options.bundle.defaultRoot`, but there are cases where `indexOf` is likely to be often `>= 0`, thus we'll end up calling both `indexOf` & `lastIndexOf`.
It simply depends on the particular input.

The default value we use is `#/definitions`, so if it's a OAS3 document that path is indeed never going to be hit, buuuut if it's OAS2 or an older JSON Schema one may use `#/definitions`, and the results might be different.
Yet again it depends on the scenario, because one may have no $refs under `#/definitions`.

In any case, feels like the best would be to make `defaultRoot` be null if one doesn't care about this output optimization.
You certainly don't, as this only affects the output document.
In other words, I'm going to expose a way to avoid this piece of code entirely.
We already have everything in place, just need a small tweak.

File: lib/pointer.js [hunk 3]
```diff
@@ -189,6 +189,25 @@ Pointer.parse = function (path, originalPath) {
...
+Pointer.parseLength = function (path) {
+  // Get the JSON pointer from the path's hash
+  let pointer = url.getHash(path).substr(1);
+
+  // If there's no pointer, then there are no tokens
+  if (!pointer) {
+    return 0;
+  }
+
+  return (pointer.match(/\//g) || []).length;
```
Comment (P0lip) [inline lib/pointer.js:208]: Since we're here for speed, I suppose that

```
let lastIndex = -1;
let i = 0;
while ((lastIndex = pointer.indexOf('/', lastIndex + 1)) !== -1) i++;
return i;
```
might be considerably faster, but if we spend, say, 4ms on `parseLength` calls then the diff would be like 1-2ms, meaning there's no point in changing anything.

**Review by P0lip — COMMENTED** [review]
Side note - I spotted an issue with our implementation, going to fix it and will add this `defaultRoot: null` too at the same time.

**Comment by marcelltoth** [comment]
This is now obsolete, #26 is a better approach 👏 

Hey @P0lip 👋 😎
